CompareFASTA_or_FASTQ utilities
===============================

Repo for my own computational resources dealing with comparing sequences in FASTA or FASTQ format (both nucleic and protein), plus give me a place to reference other handy resources.

- `compare_organisms_in_two_files_of_fasta_entries.py`

  See my comparing lists resources for related items.

- `roughly_score_relationships_to_subject_seq_pairwise_premsa.py`

  This script takes sequences in FASTA format (single multi-FASTA file) and makes quick assessment of similarity of first sequence to each of the others. It makes a matrix of the differences.

  This is located [here](https://github.com/fomightez/sequencework/tree/master/alignment-utilities) in my [alignment-utlities sub-repo](https://github.com/fomightez/sequencework/tree/master/alignment-utilities).

  This script should probably be here in this sub-repo; however, I was thinking about it in context of comparing sequences in a rough manner prior to making a full sequence alignment and put it there.


Related utilities in my other repositories
------------------------------------------

- `roughly_score_relationships_to_subject_seq_pairwise_premsa.py`

  This script takes sequences in FASTA format (single multi-FASTA file) and makes quick assessment of similarity of first sequence to each of the others. It makes a matrix of the differences.

  This is located [here](https://github.com/fomightez/sequencework/tree/master/alignment-utilities) in my [alignment-utlities sub-repo](https://github.com/fomightez/sequencework/tree/master/alignment-utilities).

  This script should probably be here in this sub-repo; however, I was thinking about it in context of comparing sequences in a rough manner prior to making a full sequence alignment and put it there.
  
