sequencework
============

Mainly python scripts related to nucleic or protein sequence work

I sorely need to put an index here with links to better guide to the appropriate folders. <== **TO DO**  
(For now look at the title of the folders to try and discern if it is something of interest.)

Descriptions of the scripts are found within `README.md` files in the sub folders.


Related 'Binderized' Utilities
----------------------------

Collection of links to launchable Jupyter environments where various sequence analysis tools work **WITHOUT ANY NEED FOR ADDITIONAL EFFORT/INSTALLS**. Many of my recent scripts are built with use in these environments in mind:

(Many of these include/feature Biopython, too, such as  but I haven't made a one all encompassing one yet for that since I use it a lot as an underlying library.)

- [patmatch-binder](https://github.com/fomightez/patmatch-binder) - launchable Jupyter sessions for running command line-based PatMatch in Jupyter environment provided via Binder (*Perl and Python-based*).

- [blast-binder](https://github.com/fomightez/blast-binder) - launchable Jupyter sessions for running command line-based BLAST+ in Jupyter environment provided via Binder.

- [InterMine-binder](https://github.com/fomightez/intermine-binder) - 
Intermine Web Services available in a Jupyter environment running via the Binder service. (See the [ guide to getting started with using Intermine sites and Jupyter using MyBinder-served Jupyter notebooks](https://github.com/fomightez/guide_to_intermine-binder).)

- [mcscan-binder](https://github.com/fomightez/mcscan-binder) - 
MCscan software available in a launchable Jupyter environment running via the Binder service (*Python 2-based*), with an example workflow and some other use examples.

- [mcscan-blast-binder](https://github.com/fomightez/mcscan-blast-binder) - MCscan and BLAST+ command line software available in a launchable Jupyter environment running via the Binder service (*Python 2-based*).

- [synchro-binder](https://github.com/fomightez/synchro-binder) - SynChro software available in a launchable Jupyter environment running via the Binder service with Quick start and some other illustrations of its use.

- [cl_sq_demo-binder](https://github.com/fomightez/cl_sq_demo-binder) - launchable, working Jupyter-based environment that has a collection of demonstrations of useful resources on command line (or useable in Jupyter sessions) for manipulating sequence files. (Note: THIS WAS STARTED AFTER SEVERAL OTHER DEMO NOTEBOOKS (many meant to be static) MADE FOR SEQUENCE SCRIPTs, and hopefully slowly those will be added to here as well to be available in active form.)

- [clausen_ribonucleotides binder](https://github.com/fomightez/clausen_ribonucleotides) - Analyze ribonucleotide incorporation data from Clausen et al. 2015 data using script `plot_5prime_end_from_bedgraph.py`.


- [circos-binder](https://github.com/fomightez/circos-binder) - Circos software available in a launchable Jupyter environment running via the Binder service with tutorials illustrating use (TBD)(*Perl and Python-based*).

Related resources by others
---------------------------

- [genomepy](https://github.com/vanheeringen-lab/genomepy)
>"Install and use genomes & gene annotations the easy way!  
genomepy is designed to provide a simple and straightforward way to download and use genomic data. This includes (1) searching available data, (2) showing the available metadata, (3) automatically downloading, preprocessing and matching data and (4) generating optional aligner indexes. All with sensible, yet controllable defaults. Currently, genomepy supports UCSC, Ensembl and NCBI." - Includes an S. cerevisiae example.
- [rna-tools (previously rna-pdb-tools): a toolbox to analyze sequences, structures and simulations of RNA](https://github.com/mmagnus/rna-tools/blob/master/index-of-tools.md)
- ['seqrequester', a tool for summarizing, extracting, generating and modifying DNA sequences.](https://github.com/marbl/seqrequester). Perl-based.
- [SeqKit - a cross-platform and ultrafast toolkit for FASTA/Q file manipulation](https://github.com/shenwei356/seqkit) by shenwei356 has some excellent utilities for handling FASTA or FASTQ files. See [here](https://github.com/shenwei356/seqkit#subcommands) for subcommands listing. [Documentation](https://bioinf.shenwei.me/seqkit/).

See also
-------

[My simulated data repo](https://github.com/fomightez/simulated_data) has some useful scripts and resources for generating simulated (mock / fake) sequence data, gene expression data, or gene lists.





